File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.py
import sys

for test_case in range(1):
    n = int(sys.stdin.readline())
    graph = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]

    #방향 0 => 오른쪽 1 => 아래쪽, 2=> 대각선


    dp = [[[0 for _ in range(n)] for _ in range(n)] for i in range(3)]
    dp[0][0][1] = 1 # 처음 위치 초기화
    for j in range(2, n):
        if graph[0][j] == 0:
            dp[0][0][j] = dp[0][0][j-1]


    for i in range(1, n):
        for j in range(1, n):
            if graph[i][j] == 0 and graph[i - 1][j] == 0 and graph[i][j - 1] == 0:
                dp[2][i][j] = dp[0][i - 1][j - 1] + dp[1][i - 1][j - 1] + dp[2][i - 1][j - 1]
            if graph[i][j] == 0:
                dp[0][i][j] = dp[0][i][j - 1] + dp[2][i][j - 1]  # 가로
                dp[1][i][j] = dp[1][i - 1][j] + dp[2][i - 1][j]  # 세로

    print(sum([dp[i][n - 1][n - 1] for i in range(3)]))

File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.java

import java.util.*;
import java.util.stream.Collectors;
import java.io.*;

public class Main
{	
	static int[][] graph;
	public static void main(String args[]) throws Exception
	{
		//BufferedReader br = new BufferedReader(new FileReader("./input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));	
		
		int n = Integer.parseInt(br.readLine());
		graph = new int[n][n];
		
		for (int row = 0; row < n; row++) {
			int[] colList = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
			
			for (int col = 0; col < n; col++) {
				graph[row][col] = colList[col];
			}
		}
		
		int[][][] dp = new int[3][n + 1][n + 1];
		
		dp[0][1][2] = 1;
		for (int row = 1; row < n + 1; row++) {
			for (int col = 3; col < n + 1; col++) {
				if ((row == 1 && col == 1) || graph[row-1][col-1] == 1) continue;
				
				dp[0][row][col] = dp[2][row][col - 1] + dp[0][row][col - 1];
				dp[1][row][col] = dp[2][row - 1][col] + dp[1][row - 1][col];
				if (row >= 2 && graph[row - 2][col-1] == 0 && graph[row-1][col - 2] == 0) {
					dp[2][row][col] = dp[2][row - 1][col - 1] + dp[0][row - 1][col - 1] + dp[1][row - 1][col - 1];
				}
			}
		}
		int ans = 0;
		for (int i = 0; i< 3; i++) ans += dp[i][n][n];
		System.out.println(ans);
	}
}